cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -4,123 +4,164 @@
-                                                  config.getString(KafkaCruiseControlConfig.CLIENT_ID_CONFIG),
-                                                  DEFAULT_MAX_IN_FLIGHT_REQUEST,
-                                                  config.getLong(KafkaCruiseControlConfig.RECONNECT_BACKOFF_MS_CONFIG),
-                                                  config.getLong(KafkaCruiseControlConfig.RECONNECT_BACKOFF_MS_CONFIG),
-                                                  config.getInt(KafkaCruiseControlConfig.SEND_BUFFER_CONFIG),
-                                                  config.getInt(KafkaCruiseControlConfig.RECEIVE_BUFFER_CONFIG),
-                                                  config.getInt(KafkaCruiseControlConfig.REQUEST_TIMEOUT_MS_CONFIG),
-                                                  true,
-                                                  new ApiVersions());
     _metadataTTL = metadataTTL;
-    // This is a super confusing interface in the Metadata. If we don't set this to false, the metadata.update()
-    // will remove all the topics that are not in the metadata interested topics list.
-    _metadata.addListener((cluster, unavailableTopics) -> _metadata.needMetadataForAllTopics(false));
+    _version = 0;
+    _cluster = Cluster.empty();
   }
 
   /**
    * Refresh the metadata. The method is synchronized because the network client is not thread safe.
    */
-  public synchronized ClusterAndGeneration refreshMetadata() {
+  public ClusterAndGeneration refreshMetadata() {
     return refreshMetadata(_refreshMetadataTimeout);
   }
 
   /**
-   * Refresh the metadata. The method is synchronized because the network client is not thread safe.
+   * Refresh the metadata. Synchronized to prevent concurrent updates to the metadata cache
    */
-  public synchronized ClusterAndGeneration refreshMetadata(long timeout) {
+  public synchronized ClusterAndGeneration refreshMetadata(int timeout) {
     // Do not update metadata if the metadata has just been refreshed.
-    if (_time.milliseconds() >= _metadata.lastSuccessfulUpdate() + _metadataTTL) {
-      // Cruise Control always fetch metadata for all the topics.
-      _metadata.needMetadataForAllTopics(true);
-      int version = _metadata.requestUpdate();
-      long remaining = timeout;
-      Cluster beforeUpdate = _metadata.fetch();
-      boolean isMetadataUpdated = _metadata.version() > version;
-      while (!isMetadataUpdated && remaining > 0) {
-        _metadata.requestUpdate();
-        long start = _time.milliseconds();
-        _networkClient.poll(remaining, start);
-        remaining -= (_time.milliseconds() - start);
-        isMetadataUpdated = _metadata.version() > version;
+    if (_time.milliseconds() >= _lastSuccessfulUpdate + _metadataTTL) {
+      // Cruise Control always fetches metadata for all the topics.
+      boolean isMetadataUpdated = false;
+      Optional<Cluster> refreshedCluster = Optional.empty();
```

Comment:
I added the check for redundancy. I went with `Optional` because I generally prefer it as it avoids NPEs. I don't think it's a big deal when the usage literally a couple of lines below - I'll change to null

Reply:
In fact, I consolidated the if block into the try